Fix 422: "Display duration of time range in the breadcrumbs bar"

MarcelloNicoletti opened the pull request:
This fixes #422.

I worked on this in collarboration with @mojac

We added a new item to the `FilterNavigatorBar` that conditionally shows when there is a selection in the profile viewer. This new element reflects the duration of the selection persistantly in addition to the current one which hides when the seleciton is not being edited.

We aren't sure what kind of styling we should apply to this new item. As it is now, we did not create any new CSS classes, electing to reuse existing ones. In order to be able to add the new item conditionally we wrapped the original `<ol>` in a surrounding `<div>` with the same classes. This means that the new `<div>` flexes to the full width *and* the `<ol>` inside it does as well. This puts our newly added item all the way to the right of the free space, right next to the "Share..." button.

<img width="808" alt="demo" src="https://user-images.githubusercontent.com/1910383/39029110-5fba5730-4417-11e8-9147-b472de3c8470.png">

julienw commented on src/components/app/ProfileFilterNavigator.js:
```diff
@@ -21,10 +21,12 @@ type StateProps = $ReadOnly<$Exact<$Diff<Props, DispatchProps>>>;
 const options: ExplicitConnectOptions<{||}, StateProps, DispatchProps> = {
   mapStateToProps: state => {
     const items = getRangeFilterLabels(state);
+    const profileSelection = state.profileView.viewOptions.selection;
```
To get the selection, we usually use selectors. We do have one here: https://github.com/devtools-html/perf.html/blob/54f197d7a100cedfb9a9f3d0d7136648d92ed6da/src/reducers/profile-view.js#L526-L527

Then I think you can calculate the selection duration here, and pass it directly as an "uncommitedItem" as was suggested by Greg in #422, so that `FilterNavigatorBar` stays generic.

julienw commented on src/components/shared/FilterNavigatorBar.js:
```diff
@@ -7,13 +7,17 @@
   +items: string[],
   +onPop: number => *,
   +selectedItem: number,
+  +profileSelection: ProfileSelection,
```
As said above, I believe this should be an optional string `uncommitedItem`, so you would write this like this:
```javascript
type Props = {|
  ...
  +uncommitedItem ?: string,
|};

julienw commented on src/components/shared/FilterNavigatorBar.js:
```diff
@@ -29,34 +33,51 @@ class FilterNavigatorBar extends PureComponent<Props> {
-              })}
-              title={item}
-              onClick={this._onLiClick}
+      <div className={classNames('filterNavigatorBar', className)}>
```
We don't need this `div` if you put the `uncommitedItem` directly inside the `<ol>`.

julienw commented on src/components/shared/FilterNavigatorBar.js:
```diff
@@ -29,34 +33,51 @@ class FilterNavigatorBar extends PureComponent<Props> {
+          ))}
+        </TransitionGroup>
+        {profileSelection.hasSelection &&
+        className === 'profileFilterNavigator' ? (
```
We don't need to check the `className`, just the existence of `uncommitedItem`. So that we don't have an hardcoded check :)

julienw commented on src/components/shared/FilterNavigatorBar.js:
```diff
@@ -29,34 +33,51 @@ class FilterNavigatorBar extends PureComponent<Props> {
+            <span className="filterNavigatorBarItemContent">
+              {getFormattedTimeLength(
+                profileSelection.selectionEnd - profileSelection.selectionStart
+              )}
```
As said above, this calculation should directly go to `ProfileFilterNavigator`. Here we should only display the `uncommitedItem` if present. No CSSTransition for this item ! :-)

Having a new class name would be a good idea too, so that we can add an opacity to make it appear paler.

julienw commented on src/components/shared/TransformNavigator.js:
```diff
@@ -24,10 +24,12 @@ type StateProps = $Diff<Props, DispatchProps>;
       className: 'calltreeTransformNavigator',
       items,
       selectedItem: items.length - 1,
+      profileSelection: profileSelection,
```
As said above, the `uncommitedItem` will be optional so you don't need to change anything here

MarcelloNicoletti commented:
@julienw I did see that suggestion by Greg in #422. It looked to me like he was suggesting adding a `<div>` after the `<ol>` and wrapping the whole thing in a fresh `<div>`. He also said:

> I would do it outside the normal structure of the `<li>` elements, and only show it if there is a sub selection.

Which is where I got the idea to add it outside of that. I can work on these fixes a bit later today.

One thing I just thought about if the `uncommittedItem` is placed right next to (or within) the current breadcrumbs and the user accepts it the `uncommittedItem` would disappear without animation while the new item would animate in. If so maybe we could instead animate it's opacity from pale to solid? I'll have to try it out and see how it looks with the current animation before I judge it though

As this is somewhat for class work and finals are coming up I wont be able to start on the test right away, but I am willing to get back to it once I get more time to do so!

MarcelloNicoletti commented on src/components/app/ProfileFilterNavigator.js:
```diff
@@ -28,7 +28,7 @@ const options: ExplicitConnectOptions<{||}, StateProps, DispatchProps> = {
           profileSelection.selectionEnd - profileSelection.selectionStart
         )
-      : undefined;
+      : ''; // Used to be undefined, but it seems Redux won't merge that
```
This is less than ideal. It adds an empty list item at all times rather than not having one when there is no selection

julienw commented:
> If so maybe we could instead animate it's opacity from pale to solid? I'll have to try it out and see how it looks with the current animation before I judge it though

Yes I think that's a good idea, if you can try it !

> For example, setting it to 'undefined' inside mapStateToProps does update the item when unsetting the selection.

I guess you mean `doesn't update` here, right ? That's a good finding, I wasn't aware of this. As a workaround I can suggest using `null` or, as you did, `''`.

(More comments coming with a review)

julienw commented on src/components/shared/FilterNavigatorBar.css:
```diff
@@ -102,6 +102,14 @@
 
+.filterNavigatorBarUncommittedItem {
+  opacity: 0.75;
+}
```
I think the opacity is still too high; can you try with 0.65 ?

Also, like we discussed in a separate commit, I think we can try to animate going from this class to the commited value. I think we can animate on both `opacity` and `color`. Please do this in a separate commit :)

julienw commented:
Oh yeah, you're totally right. Please keep the `?` then :)

Le mar. 24 avr. 2018 21:23, Marcello Nicoletti <notifications@github.com> a
écrit :

> *@MarcelloNicoletti* commented on this pull request.
> ------------------------------
>
> In src/components/shared/FilterNavigatorBar.js
> <https://github.com/devtools-html/perf.html/pull/952#discussion_r183851251>
> :
>
> > @@ -14,6 +14,7 @@ type Props = {|
>    +items: string[],
>    +onPop: number => *,
>    +selectedItem: number,
> +  +uncommittedItem?: string,
>
> The TranformNavigator also uses this FilterNavigatorBar. Right now we
> aren't passing a string in from there, instead it is undefined. Should I
> add an empty string there and remove the ? or should I leave it as is?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/devtools-html/perf.html/pull/952#discussion_r183851251>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAbuH0qXVllnmU6tGpowp_9ldbssrbOZks5tr3vFgaJpZM4Tcz5i>
> .
>

MarcelloNicoletti commented:
Hm. Chaning the condition from `uncommittedItem !== undefined ? [...]` to `uncommittedItem ? [...]` actually causes the same behavior I was seeing with `undefined` before using the empty string. I'm not sure if they're caused by the same issue or different issues. With the name itself as the condition the old value sticks around rather than disappearing when `mapStateToProps` sets `uncommittedItem` to the empty string.

This caused me to experiment more than I had earlier. I changed `null` to be a dummy `<li>` component so that when the condition was true it returns one child and when it's false it returns the dummy. What I saw was that the empty string in `mapStateToProps` causes that dummy to show up unless there is a selection. I also changed the empty string back to `undefined`. I can now see that Redux *is* merging `undefined` and I know this because the dummy text was `{typeof uncommittedItem}` and it showed "undefined".

What I see now is that when we return a new child item from the conditional React replaces the old one and redraws, but when we return null from the conditional Redux seems to leave the old child in place. That's even in the case were other children change and the parent component redraws.

MarcelloNicoletti commented:
By the way, I'm having some issues with `node` and `flow`. Particularly that 8 or so tests fail on `node 10.0.0` due to symbolication issues. Also, after I rolled my `node` back, cleaned my `yarn` packages, and did `yarn install` again `yarn flow` reports
```
The flow server's version didn't match the client's, so it exited.
Going to launch a new one.
```
over and over again until I stop it. So for now I can't push my changes because the tests fail and `huskey` stops pushes.

julienw commented:
About node v10: I reproduce the issue, I filed #967.
About flow: try running `yarn flow stop`. If this still doesn't work, look at your processes, sometimes some flow processes are still there... and manually kill them. If this still doesn't work, you can still push with `--no-verify`.

About your problem, I investigaged locally and I found that this actually comes from the TransitionGroup. From [the documentation](https://reactcommunity.org/react-transition-group/) its children needs to be `Transition` elements... We initially thought that this wasn't needed, but this is needed. By the way there are errors in the console because of this.

Here we want a transition that's not a transition, so I tried it with a `<Transition timeout={0}>` (`timeout` is mandatory) and it worked properly. This also means you can pass `undefined` instead of `''` :)

But... actually as said before we'll want a `exit` transition: so maybe what we really want is a `CSSTransition` so that we can apply some CSS properties. But maybe that's not needed, I'm not sure.
Another tip is that you may want to add a `key={items.length}` to the Transition, so that when we commit the new range, the "temporary" selection blends in to the new commited selection (the key would be equal to the "next" committed element). As said before, this should be for a separate commit, and if you prefer this can even be done in a separate pull request.

Note that I'll be in holidays until wednesday next week and I won't look at my mails. I think @zoepage or @gregtatum will be able to help you finish this PR if needed and will be able to give you the final review. We're nearly there !

julienw commented on src/components/shared/FilterNavigatorBar.js:
```diff
@@ -14,6 +14,7 @@ type Props = {|
   +items: string[],
   +onPop: number => *,
   +selectedItem: number,
+  +uncommittedItem?: string,
```
You're totally right, please leave it as is!

julienw commented:
Hey @MarcelloNicoletti, how are you? Do you still intend to finish this PR up?
If not the team can finish this on your behalf and commit while keeping your author information. Just tell us :-)

Thanks!

julienw commented:
> Yes, I do intend to finish this up. Sorry for the delay. I had finals starting May 3rd and I graduated on May 11th. I’m finally free to start back up on it again today!

Congrats ! And I'm glad you're still with us :-) Please ping us if you need anything

MarcelloNicoletti commented:
My flow issue has been fixed and I was able to push. I think I've implemented what I could without working on the animation or test. Should I add the test to this PR or make a new one later on?

gregtatum commented:
Friendly ping, @julienw it looks like this PR is waiting on your feedback.
